src/Hl7.Fhir.Core/Rest/BaseFhirClient.cs

```diff
@@ -0,0 +1,1025 @@
+﻿using Hl7.Fhir.Model;
+using Hl7.Fhir.Specification;
+using Hl7.Fhir.Utility;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Hl7.Fhir.Rest
+{
+    public abstract partial class BaseFhirClient : IDisposable
+    {
+        protected BaseFhirClient(Uri endpoint, FhirClientSettings settings = null)
+        {
+            Settings = (settings ?? new FhirClientSettings());
+            Endpoint = getValidatedEndpoint(endpoint);
+        }
+
+        protected IRequester Requester { get; set; }
+
+        /// <summary>
+        /// The default endpoint for use with operations that use discrete id/version parameters
+        /// instead of explicit uri endpoints. This will always have a trailing "/"
+        /// </summary>
+        public Uri Endpoint
+        {
+            get;
+            protected set;
+        }
+
+        public FhirClientSettings Settings = new FhirClientSettings();
+
+        /// <summary>
+        /// The last transaction result that was executed on this connection to the FHIR server
+        /// </summary>
+        public Bundle.ResponseComponent LastResult { get; private set; }
+
+        public byte[] LastBody => LastResult?.GetBody();
+        public string LastBodyAsText => LastResult?.GetBodyAsText();
+        public Resource LastBodyAsResource { get; private set; }
+
+        protected object LastClientRequest { get; set; }
+        protected object LastClientResponse { get; set; }
+
+        private static Uri getValidatedEndpoint(Uri endpoint)
+        {
+            if (endpoint == null) throw new ArgumentNullException("endpoint");
+
+            if (!endpoint.OriginalString.EndsWith("/"))
```

Comment:
I think we have to create an helper function for this, because I see it's been used in a lot of places.
```
public static string EnsureEndsWith(this string original, string toEndWith)
{
     original = original ?? string.Empty;
     return original + (original.EndsWith(toEndWith) ? string.Empty : toEndWith);
}
```

This should go to `Hl7.Fhir.Utility.StringExtensions` (Common)

Reply:
done